kudu-handler/src/java/org/apache/hadoop/hive/kudu/KuduPredicateHandler.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPLessThan;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPNotNull;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPNull;
+import org.apache.kudu.ColumnSchema;
+import org.apache.kudu.Schema;
+import org.apache.kudu.client.KuduPredicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Contains static methods for decomposing predicate/filter expressions and
+ * getting the equivalent Kudu predicates.
+ */
+public final class KuduPredicateHandler {
+  static final Logger LOG = LoggerFactory.getLogger(KuduPredicateHandler.class);
+
+  private KuduPredicateHandler() {}
+
+  /**
+   * Analyzes the predicates and return the portion of it which
+   * cannot be evaluated by Kudu during table access.
+   *
+   * @param predicateExpr predicate to be decomposed
+   * @param schema the schema of the Kudu table
+   * @return decomposed form of predicate, or null if no pushdown is possible at all
+   */
+  public static DecomposedPredicate decompose(ExprNodeDesc predicateExpr, Schema schema) {
+    IndexPredicateAnalyzer analyzer = newAnalyzer(schema);
+    List<IndexSearchCondition> sConditions = new ArrayList<>();
+    ExprNodeDesc residualPredicate = analyzer.analyzePredicate(predicateExpr, sConditions);
+
+    // Nothing to decompose.
+    if (sConditions.size() == 0) {
+      return null;
+    }
+
+    DecomposedPredicate decomposedPredicate = new DecomposedPredicate();
+    decomposedPredicate.pushedPredicate = analyzer.translateSearchConditions(sConditions);
+    decomposedPredicate.residualPredicate = (ExprNodeGenericFuncDesc) residualPredicate;
+    return decomposedPredicate;
+  }
+
+  /**
+   * Returns the list of Kudu predicates from the passed configuration.
+   *
+   * @param conf the execution configuration
+   * @param schema the schema of the Kudu table
+   * @return the list of Kudu predicates
+   */
+  public static List<KuduPredicate> getPredicates(Configuration conf, Schema schema) {
+    List<KuduPredicate> predicates = new ArrayList<>();
+    for (IndexSearchCondition sc : getSearchConditions(conf, schema)) {
+      predicates.add(conditionToPredicate(sc, schema));
+    }
+    return predicates;
+  }
+
+  private static List<IndexSearchCondition> getSearchConditions(Configuration conf, Schema schema) {
+    List<IndexSearchCondition> conditions = new ArrayList<>();
+    ExprNodeDesc filterExpr = getExpression(conf);
+    if (null == filterExpr) {
+      return conditions;
+    }
+    IndexPredicateAnalyzer analyzer = newAnalyzer(schema);
+    ExprNodeDesc residual = analyzer.analyzePredicate(filterExpr, conditions);
+
+    // There should be no residual since we already negotiated that earlier in
+    // decomposePredicate. However, with hive.optimize.index.filter
+    // OpProcFactory#pushFilterToStorageHandler pushes the original filter back down again.
```

Comment:
Can we open a bug for this to keep track?

Reply:
It looks like this is tracked by HIVE-6678. I will add the jira to the comment.